### ISSUE TITLE
3D particles are not visible through transparent renderables

### ISSUE THREAD

**Issue by JackRainy** [libgdx/libgdx#6575]
#### Issue details
This issue looks similar to #3742 , however, I can't confirm that for sure.
The problem is that PointSprite particles sometimes depending on camera angle are not visible behind the transparent models.

#### Reproduction steps/code
I have created a simple [ParticleTransparencyTest](https://github.com/JackRainy/libgdx/blob/particlesTransparency/tests/gdx-tests/src/com/badlogic/gdx/tests/g3d/ParticleTransparencyTest.java) to illustrate that.
The source code is in my branch of the fork: https://github.com/JackRainy/libgdx/tree/particlesTransparency
![Screenshot 2021-06-24 01-36-25](https://user-images.githubusercontent.com/27405436/123176991-1e5fbc80-d48d-11eb-85b7-55359f947477.png)
As you can see on these examples: the far box is always non-transparent for the far particles. But the closer box is transparent for both middle and far particles.
On the left screen: the red particles are covered. On the right screen (after rotating the camera 180°): the blue particles are covered now.

Beside this, I have tried to rebase and apply PR #4318 . Probably, I did some mistakes resolving conflicts during rebase, but the PR does not fix the issue. Also, the ParticleSystemBlendingTest (provided by the PR's author) fails as well.
![image](https://user-images.githubusercontent.com/27405436/123177994-00935700-d48f-11eb-9cf5-f46456137c46.png)
You can see that red and blue particles closer in the center are also trimmed by transparent boxes, while the far red one is visible through it.

#### Version of LibGDX and/or relevant dependencies
version 1.10.0

#### Please select the affected platforms
I have seen this issue on both Windows and Android.

**Comment by mgsx-dev** [comment]
@JackRainy Hi, i rebased the PR #4318, i think it's OK now. It should fix your issue, make sure to enable splitRenderablesPerController for your pointSpriteParticleBatch.

However the issue you pointed out about particles trimmed when rendered on top of transparent box is a known limitation. There's nothing we can do about it, it would require to merge both models and particles triangles and sort them CPU side before rendering. That would require a full rewrite of libgdx 3D batching and may have a lot of performance impacts.

The only workaround is to avoid transparent things inside or very close to other transparent things.

**Comment by JackRainy** [comment]
Thank you for the prompt feedback. 
I just took the latest rebased #4318 and it is still not working as I could expect.
| splitRenderablesPerController = false | splitRenderablesPerController = true |
|---|---|
|  ![Screenshot 2021-06-24 21-35-46](https://user-images.githubusercontent.com/27405436/123315792-c8deea80-d534-11eb-9c94-79d88ecf98d5.png) | ![Screenshot 2021-06-24 21-36-19](https://user-images.githubusercontent.com/27405436/123315803-caa8ae00-d534-11eb-86bb-9c2c44652fc2.png) |

Probably, it's the same trimming problem you mentioned. The trimming happens in different areas with regard to the option on/off.

>  it would require to merge both models and particles triangles and sort them CPU side before rendering.

Beside this, I faced the issue with Decals. I assume the problem has the same origin.
There is a question on Stackoverflow regarding that: https://stackoverflow.com/questions/32287895/libgdx-culling-between-modelinstances-decals
I can't get Decals and Models sorted in one common batch, so the workaround for me is just not to use Decals at all but replacing them with other Renderables compatible with ModelBatch.

**Comment by mgsx-dev** [comment]
Yeah, that's the same problem. When rendering transparent things, they must be sorted back to front. When mixing particles, decals, and models, this can be hard to do.
For particles, this new splitRenderablesPerController option produce 1 renderable (or more) per particles group to be sorted by model batch. That works for most of cases. In your last screenshot, you can see on the left that it's mainly broken because all particles are rendered in one draw call. On the right, it's only broken for some edge cases (particles inside cubes). The test class includes both working cases and limited cases.

For decals, you need to group them as well, for instance you could render far ones (flushing DecalBatch), then render transparent models (flushing ModelBatch), then render front decals.
If things overlaps (eg. particles inside a transparent model), there is no acceptable solution.

That said, i think my PR fix your original scene. Is that the case?

**Comment by JackRainy** [comment]
Yes, thank you, @mgsx-dev, for the initial case, having the PR applied with `splitRenderablesPerController = true` it works like a charm!
![image](https://user-images.githubusercontent.com/27405436/123325270-7e636b00-d540-11eb-8166-167f9199f5e0.png)
